driver-bitfury16.c

```diff
@@ -0,0 +1,4589 @@
+	if ((info->chipboard_num == 0) ||
+		(info->active_chipboard_num == 0)) {
+		deinit_x5(bitfury);
+
+		/* close devices */
+		close_spi_device(SPI_CHANNEL1);
+		close_spi_device(SPI_CHANNEL2);
+		close_ctrl_device();
+		close_uart_device(UART_CHANNEL1);
+		close_uart_device(UART_CHANNEL2);
+
+#ifdef FILELOG
+		fclose(info->logfile);
+#endif
+
+		applog(LOG_ERR, "%s: no boards present. exiting...",
+				bitfury->drv->name);
+
+		free(info);
+		free(bitfury);
+
+		return;
+	}
+
+	mutex_init(&info->nonces_good_lock);
+
+	if (!add_cgpu(bitfury))
+		quit(1, "%s: %s() failed to add_cgpu",
+				bitfury->drv->name, __func__);
+
+	info->initialised = true;
+
+	applog(LOG_INFO, "%s: chip driver initialized", bitfury->drv->name);
+#ifdef FILELOG
+	filelog(info, "%s: cgminer started", bitfury->drv->name);
+#endif
+}
+
+static uint8_t chip_task_update(struct cgpu_info *bitfury, bf_chip_address_t chip_address)
+{
+	uint8_t i;
+	uint8_t ret = 0;
+	bf_works_t work;
+	time_t curr_time_t;
+	struct timeval curr_time;
+
+	uint8_t board_id  = chip_address.board_id;
+	uint8_t bcm250_id = chip_address.bcm250_id;
+	uint8_t chip_id   = chip_address.chip_id;
+
+	struct bitfury_info *info = (struct bitfury_info *)(bitfury->device_data);
+	bf_cmd_buffer_t* cmd_buffer = &info->chipboard[board_id].cmd_buffer;
+
+	if (cmd_buffer->status != EMPTY)
+		return -1;
+
+	switch (info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].status) {
+
+	/* fill chip buffer with toggle task */
+	case UNINITIALIZED:
+		applog(LOG_DEBUG, "%s: chipworker_thr: chip [%d:%d:%2d], prepare toggle_cmd",
+				bitfury->drv->name,
+				board_id, bcm250_id, chip_id);
+
+		uint8_t toggle[4] = { 0xa5, 0x00, 0x00, 0x02 };
+
+		ret = cmd_buffer_push(cmd_buffer,
+				info->chipboard[board_id].bcm250[bcm250_id].channel_depth, chip_address, chip_address,
+				work, 0, CHIP_CMD_TOGGLE, 3, toggle);
+
+		if (ret < 0)
+			applog(LOG_ERR, "%s: chipworker_thr: chip [%d:%d:%2d], error prepare toggle_cmd",
+					bitfury->drv->name,
+					board_id, bcm250_id, chip_id);
+		break;
+
+	/* fill chip buffer with set clock task */
+	case TOGGLE_SET:
+		applog(LOG_DEBUG, "%s: chipworker_thr: chip [%d:%d:%2d], prepare set_clock_cmd",
+				bitfury->drv->name,
+				board_id, bcm250_id, chip_id);
+
+		uint8_t clock_buf[4];
+		memset(clock_buf, 0, sizeof(clock_buf));
+
+		/* init renonce chip with lower clock */
+		if ((renonce_chip(chip_address) == 1) &&
+			(opt_bf16_renonce != RENONCE_DISABLED)) {
+			gen_clock_data(bf16_renonce_chip_clock, 1, clock_buf);
+			info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].clock = bf16_renonce_chip_clock;
+		} else {
+			gen_clock_data(bf16_chip_clock, 1, clock_buf);
+			info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].clock = bf16_chip_clock;
+		}
+
+		ret = cmd_buffer_push(cmd_buffer,
+				info->chipboard[board_id].bcm250[bcm250_id].channel_depth, chip_address, chip_address,
+				work, 0, CHIP_CMD_SET_CLOCK, 3, clock_buf);
+
+		if (ret < 0)
```

Comment:
driver-bitfury16.c:1524:11: warning: comparison is always false due to limited range of data type [-Wtype-limits]
   if (ret < 0)